### ISSUE TITLE
onBeforeAction runs before data

### ISSUE THREAD

**Issue by zimt28** [iron-meteor/iron-router#1247]
In iron-router versions prior to 1 the `onBeforeAction` did run before `data`, which was helpful in some cases. Now that has changed, `data` runs before `onBeforeAction`. Why has that changed? Is it intended?

**Comment by victorbstan** [comment]
I've had issues with onBeforeAction, I don't even know if it's still supported for sure. I switched to using `onRun`. 

For example: 

```
Router.onRun(requireLogin, { only: [
  'postSubmit',
  'dashboard',
  'locations',
  'properties',
  'providers'
] });
```

**Comment by zimt28** [comment]
Thank you! `onRun` works for me, too. The `onBeforeAction` still seems to be supported, as I can find it in the guide. It would be nice if it would just work like it did before. It might be helpful in cases like this:

I'm using the [space:ui package](https://github.com/CodeAdventure/space-ui) and need to tell the store, which document I want. In `iron:router` < v1, the following worked great, the router checked if the document exists: 

```coffeescript

class App.CasesPrintController

  Dependencies:
    actions: 'Actions'
    meteor: 'Meteor'
    router: 'Router'
    store: 'CasesStore'
    tracker: 'Tracker'

  onDependenciesReady: ->

    self = this

    @router.route 'cases.print',
      path: '/print/:_id'

      waitOn: ->
        singleSubs.subscribe 'printCase', @params._id

      data: ->
        self.store.getState().case.fetch()[0]

      onBeforeAction: ->
        caseId = @params._id

        self.tracker.nonreactive -> self._setCurrentCase caseId

        @next()

      action: ->
        unless self.meteor.user()
          @layout 'PrintLayout'

        @render()  if @ready()

  _setCurrentCase: (caseId) ->
    @actions.setCurrentCase caseId
```

Now, as `onBeforeAction` runs after `data`, I have to put all the stuff from the `onBeforeAction` into `data`, which seems hacky.

I also think this behaviour might explain my issue here: https://github.com/zimme/meteor-iron-router-auth/issues/9

**Comment by victorbstan** [comment]
Yeah, I'm not sure what's going on here, I was expecting onBeforeAction to work as well, I just noticed it didn't yesterday as well.

**Comment by zimt28** [comment]
The problem with `onRun` is, that in some cases the actual page just flashes, before the login page appears.

**Comment by zimt28** [comment]
The problem with `onRun` is, that in some cases the actual page just flashes, before the login page appears. @victorbstan do you know how to solve this?

**Comment by victorbstan** [comment]
@zimt28 

I'm not sure what you are seeing. I don't have flashes, if log out and try to visit a page where I'm checking for authentication in onRun, the visitor is redirected to the login page right away... might depend on how you have your routes/controllers set-up?

**Comment by zimt28** [comment]
I'm currently using zimme/meteor-iron-router-auth, but I'll try your solution later this day.

**Comment by sferoze** [comment]
I've noticed this too. I console.log(1) and console.log(2) on the onBeforeAction and data function and the data runs first. This is not what one would expect and it pretty frustrating. onBeforeAction should run before data.

**Comment by sferoze** [comment]
When I switched to onRun it seems to work. is onRun the first thing that is run when a route is visited?

**Comment by rista404** [comment]
Hey folks, maybe a different topic but is onRun safe for auth checking? 

@victorbstan @sferoze
